### PR TITLE
Native/builtin programs now receive an InvokeContext

### PR DESCRIPTION
**Pull request by mvines**
By plumbing InvokeContext into native and builtin programs, we open the door for:
1) Feature gating *within* the program.  🎉 
2) Enabling CPI, which is needed for unit testing

Downsides:
* FeatureSet moves into sdk/  - ok not really a downside
* `runtime/src/process_instruction.rs` → `sdk/src/process_instruction.rs` -- not great but now that `solana-program` exists, this seems fine
* `runtime/src/bpf_test_utils.rs` gets inlined into `sdk/src/process_instruction.rs`.  These aren't BPF-specific test utils at all right now

### CODE REVIEW

File: sdk/src/process_instruction.rs [hunk 1]
```diff
@@ -24,7 +24,6 @@ pub type LoaderEntrypoint = unsafe extern "C" fn(
     invoke_context: &dyn InvokeContext,
 ) -> Result<(), InstructionError>;
 
-pub type ProcessInstruction = fn(&Pubkey, &[KeyedAccount], &[u8]) -> Result<(), InstructionError>;
 pub type ProcessInstructionWithContext =
```
Comment (jackcmay) [inline sdk/src/process_instruction.rs:27]: Why not just call this `ProcessInstruction` now since there only needs to be one?

Comment (mvines) [reply sdk/src/process_instruction.rs:27]: will do!

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -10065,22 +10077,15 @@ mod tests {
...
             _data: &[u8],
-            _context: &mut dyn InvokeContext,
+            _invoke_context: &mut dyn InvokeContext,
         ) -> std::result::Result<(), InstructionError> {
             Ok(())
         }
-        assert!(!format!("{:?}", Entrypoint::Program(mock_process_instruction)).is_empty());
-        assert!(!format!("{:?}", Entrypoint::Loader(mock_ix_processor)).is_empty());
+        assert!(!format!("{:?}", Entrypoint::Program(mock_processor)).is_empty());
+        assert!(!format!("{:?}", Entrypoint::Loader(mock_processor)).is_empty());
```
Comment (jackcmay) [inline runtime/src/bank.rs:10089]: Do we need `Entrypoint::Loader` any longer?

Comment (mvines) [reply runtime/src/bank.rs:10089]: I guess only to add typing to distinguish a loader from a program.   I can try pulling it out and see what that does to the code though?

Comment (jackcmay) [reply runtime/src/bank.rs:10089]: Its originally there to handle the two entrypoints, now that there is one lets clean that up too

File: sdk/src/process_instruction.rs (hunk 1 above)
Comment (mvines) [reply sdk/src/process_instruction.rs:27]: I think there's value in keeping ProcessInstructionWithContext.
There's `solana_program::entrypoint::ProcessInstruction`, which has no context.   So keeping this as 
`ProcessInstructionWithContext` improves readability especially in files that use both types

File: runtime/src/bank.rs (hunk 2 above)
Comment (mvines) [reply runtime/src/bank.rs:10089]: @jackcmay - do you see any issue with deleting the `MessageProcessor::loaders` field entirely?   The only impact seems to be in `MessageProcessor::process_instruction`

Comment (jackcmay) [reply runtime/src/bank.rs:10089]: If loaders and programs have the same entrypoint signature now then we don't need to make the distinction between the two anymore.  So yeah, we can probably roll `loaders` into `programs`

Comment (mvines) [reply runtime/src/bank.rs:10089]: I just added b7d184c759754595aa560af76d5e5585c525f659 to merge them in MessageProcessor.  Once that passes CI, I'll make the larger changes need across the code base needed to de-Entrypoint

Comment (mvines) [reply runtime/src/bank.rs]: Looks good, Entrypoint is now gone

**Pull request merged**